clang/lib/CIR/CodeGen/CIRGenExprCXX.cpp

```diff
@@ -911,3 +911,88 @@ void CIRGenFunction::buildDeleteCall(const FunctionDecl *DeleteFD,
     llvm_unreachable("NYI"); // DestroyingDeleteTag->eraseFromParent();
   }
 }
+
+static mlir::Value buildDynamicCastToNull(CIRGenFunction &CGF,
+                                          mlir::Location Loc, QualType DestTy) {
+  mlir::Type DestCIRTy = CGF.ConvertType(DestTy);
+  assert(DestCIRTy.isa<mlir::cir::PointerType>() &&
+         "result of dynamic_cast should be a ptr");
+
+  mlir::Value NullPtrValue = CGF.getBuilder().getNullPtr(DestCIRTy, Loc);
+
+  if (!DestTy->isPointerType()) {
+    /// C++ [expr.dynamic.cast]p9:
+    ///   A failed cast to reference type throws std::bad_cast
+    CGF.CGM.getCXXABI().buildBadCastCall(CGF, Loc);
+  }
+
+  return NullPtrValue;
+}
+
+mlir::Value CIRGenFunction::buildDynamicCast(Address ThisAddr,
+                                             const CXXDynamicCastExpr *DCE) {
+  auto loc = getLoc(DCE->getSourceRange());
+
+  CGM.buildExplicitCastExprType(DCE, this);
+  QualType destTy = DCE->getTypeAsWritten();
+  QualType srcTy = DCE->getSubExpr()->getType();
+
+  // C++ [expr.dynamic.cast]p7:
+  //   If T is "pointer to cv void," then the result is a pointer to the most
+  //   derived object pointed to by v.
+  bool isDynCastToVoid = destTy->isVoidPointerType();
+  QualType srcRecordTy;
+  QualType destRecordTy;
+  if (isDynCastToVoid) {
+    srcRecordTy = srcTy->getPointeeType();
+    // No DestRecordTy.
+  } else if (const PointerType *DestPTy = destTy->getAs<PointerType>()) {
+    srcRecordTy = srcTy->castAs<PointerType>()->getPointeeType();
+    destRecordTy = DestPTy->getPointeeType();
+  } else {
+    srcRecordTy = srcTy;
+    destRecordTy = destTy->castAs<ReferenceType>()->getPointeeType();
```

Comment:
Are the 3 arms of `if/else` here covered in the testcase? I have the impression just some of them are covered, but please correct me if I'm wrong. You either need to add testcases for all arms or put `llvm_unreacheable("NYI")` for the arms you currently have not implemented.

Reply:
> Are the 3 arms of if/else here covered in the testcase?

The first case covers `dynamic_cast<void *>` and it's NYI. The other 2 branches are covered by the test case: one for casting pointers and the other for casting references.

> You either need to add testcases for all arms or put llvm_unreacheable("NYI") for the arms you currently have not implemented.

At present, `dynamic_cast<void *>` is guarded at [line 974](https://github.com/llvm/clangir/pull/426/files/ba4b5aa2018824f39cb268059f2b4502d2c36431#diff-5fc4ee3f9c9e0858bc9caa8f04ed0bcfa762449f0529161cb7396ece2a0fe6ceR974) in this file. Should I put the NYI guard at here?

Reply:
Updated.

Reply:
This looks good, thanks!